dagshub/data_engine/model/datapoint.py

```diff
@@ -47,10 +47,20 @@ def __setitem__(self, key, value):
 
     def save(self):
         """
-        call save to commit changes to metadata done with
-        one ore more dictionary assignment syntax usages, example:
-         'specific_data_point[metadata-name] = value
-         specific_data_point.save()'
+        Commit metadata changes on this datapoint.
+
+        Example::
+
+            datapoint["field"] = value
+            datapoint.save()
+
+        .. note::
+            If done inside a metadata context, ``save()`` doesn't need to be called,
+            all changes will be uploaded on context exit::
+
+                with ds.metadata_context():
+                    for i, dp in ds.all():
+                        dp["number"] = i
         """
 
         # if in context block, don't _upload_metadata, it will be done at context end
```

Comment:
I think there's a subtle bug here:
1. Get datapoint 1 from a datasource
2. Modify it - implicit upload context gets created
3. Get datapoint 2
4. Modify it - keeps existing context, adds to it
5. Save datapoint 1 - it actually saves datapoint 2 as well!

Reply:
I can live with it but it's bad ACID or something

Reply:
yes, you're right, it totally works that way.
I guess it's a bit of an underexplained behavior? You'd probably expect `dp.save()` to upload only that datapoint's changes.
I'll think about changing it maybe

Reply:
Would require to separate the dict assignment inside of a context from dict assignment that's coupled with `save()`. Or maybe would require to delete it from the context somehow.

Reply:
But if we don't have an implicit context, we're throwing away the functionality that we're commenting on here (the snippet in the comment).
I think it's neat and is very convenient without really any drawbacks.
I can remove it if you really want to, but i think it's good UX.

Reply:
I guess we can still have this implicit context behavior by passing the context into `save()`, with throwaway context with immediate upload by default.

Reply:
Would you be ok with moving this to another task then? This is kind of out of scope of this PR at this point

Reply:
Sure, no problem doing this in a separate PR